File: DTensor/Tensor-Implementations/Tests/OptimizerTests/test_long_horizon_drift.cpp
/**
 * @file test_long_horizon_drift.cpp
 * @brief Long-horizon drift test: 500 steps with non-uniform gradients.
 *
 * Tests Adam, AdamW, and SGD (momentum + weight_decay) over 500 steps.
 * Generates a different random gradient each step using a simple LCG PRNG
 * that is replicated identically in the companion Python script.
 *
 * Outputs CSV files with per-step parameter values for automated comparison.
 *
 * Build & run:
 *   make run-snippet FILE=Tests/OptimizerTests/test_long_horizon_drift.cpp
 *
 * Then run the Python companion:
 *   cd Tests/OptimizerTests && python3 test_long_horizon_drift.py
 */
#include <iostream>
#include <iomanip>
#include <fstream>
#include <cmath>
#include <vector>
#include <string>

#include "TensorLib.h"

using namespace OwnTensor;

// ============================================================================
// Deterministic LCG PRNG — must match Python exactly
// ============================================================================
// Returns a float in [-1.0, 1.0) given a seed.
// Uses the Numerical Recipes LCG: next = (a * seed + c) % m
static float lcg_float(uint64_t seed) {
    // Mix the seed through a few LCG rounds for better distribution
    uint64_t state = seed;
    state = (6364136223846793005ULL * state + 1442695040888963407ULL);
    state = (6364136223846793005ULL * state + 1442695040888963407ULL);
    // Map to [-1.0, 1.0)
    uint32_t bits = static_cast<uint32_t>(state >> 33);  // top 31 bits
    return (static_cast<float>(bits) / static_cast<float>(0x7FFFFFFFU)) * 2.0f - 1.0f;
}


// Generate a gradient tensor with deterministic non-uniform values
static Tensor make_gradient(int step, int numel, bool use_gpu) {
    TensorOptions opts = TensorOptions().with_dtype(Dtype::Float32);
    Tensor grad = Tensor::zeros(Shape{{4, 4}}, opts);
    float* data = grad.data<float>();
    for (int i = 0; i < numel; ++i) {
        uint64_t seed = static_cast<uint64_t>(step) * 1000ULL + static_cast<uint64_t>(i);
        data[i] = lcg_float(seed);
    }
    if (use_gpu) {
        grad = grad.to_cuda(0);
    }
    return grad;
}


// ============================================================================
// Write parameter values to CSV at each step
// ============================================================================
static void write_csv_header(std::ofstream& f, int numel) {
    f << "step";
    for (int i = 0; i < numel; ++i) {
        f << ",p" << i;
    }
    f << "\n";
}

static void write_csv_row(std::ofstream& f, int step, const Tensor& t, int numel) {
    Tensor tc = t.to_cpu();
    const float* d = tc.data<float>();
    f << step;
    for (int i = 0; i < numel; ++i) {
        f << "," << std::fixed << std::setprecision(8) << d[i];
    }
    f << "\n";
}


// ============================================================================
// Test runners
// ============================================================================
static const int NUM_STEPS = 5000;
static const int NUMEL = 16;  // 4x4

// Initial weights: [0.1, 0.2, ..., 1.6]
static Tensor make_initial_weights(bool use_gpu) {
    TensorOptions opts = TensorOptions()
        .with_dtype(Dtype::Float32)
        .with_req_grad(true);
    Tensor W = Tensor::zeros(Shape{{4, 4}}, opts);
    float* data = W.data<float>();
    for (int i = 0; i < NUMEL; ++i) {
        data[i] = 0.1f * (i + 1);
    }
    W.set_requires_grad(true);
    if (use_gpu) {
        W = W.to_cuda(0);
        W.set_requires_grad(true);
    }
    return W;
}


void test_adam(bool use_gpu, const std::string& csv_path) {
    std::string device = use_gpu ? "GPU" : "CPU";
    std::cout << "\n=== Adam (" << device << ") — " << NUM_STEPS << " steps ===" << std::endl;

    Tensor W = make_initial_weights(use_gpu);

    float lr = 0.001f, beta1 = 0.9f, beta2 = 0.999f, eps = 1e-8f, wd = 0.01f;
    nn::Adam adam({W}, lr, beta1, beta2, eps, wd);

    std::ofstream csv(csv_path);
    write_csv_header(csv, NUMEL);
    // Write initial state as step 0
    write_csv_row(csv, 0, W, NUMEL);

    for (int step = 1; step <= NUM_STEPS; ++step) {
        Tensor grad = make_gradient(step, NUMEL, use_gpu);
        W.set_grad(grad);
        adam.step();
        write_csv_row(csv, step, W, NUMEL);
    }
    csv.close();
    std::cout << "  Wrote " << csv_path << std::endl;
}


void test_adamw(bool use_gpu, const std::string& csv_path) {
    std::string device = use_gpu ? "GPU" : "CPU";
    std::cout << "\n=== AdamW (" << device << ") — " << NUM_STEPS << " steps ===" << std::endl;

    Tensor W = make_initial_weights(use_gpu);

    float lr = 0.001f, beta1 = 0.9f, beta2 = 0.999f, eps = 1e-8f, wd = 0.01f;
    nn::AdamW adamw({W}, lr, beta1, beta2, eps, wd);

    std::ofstream csv(csv_path);
    write_csv_header(csv, NUMEL);
    write_csv_row(csv, 0, W, NUMEL);

    for (int step = 1; step <= NUM_STEPS; ++step) {
        Tensor grad = make_gradient(step, NUMEL, use_gpu);
        W.set_grad(grad);
        adamw.step();
        write_csv_row(csv, step, W, NUMEL);
    }
    csv.close();
    std::cout << "  Wrote " << csv_path << std::endl;
}


void test_sgd(bool use_gpu, const std::string& csv_path) {
    std::string device = use_gpu ? "GPU" : "CPU";
    std::cout << "\n=== SGD+Momentum+WD (" << device << ") — " << NUM_STEPS << " steps ===" << std::endl;

    Tensor W = make_initial_weights(use_gpu);

    float lr = 0.01f, momentum = 0.9f, wd = 0.01f;
    nn::SGDOptimizer sgd({W}, lr, momentum, wd);

    std::ofstream csv(csv_path);
    write_csv_header(csv, NUMEL);
    write_csv_row(csv, 0, W, NUMEL);

    for (int step = 1; step <= NUM_STEPS; ++step) {
        Tensor grad = make_gradient(step, NUMEL, use_gpu);
        W.set_grad(grad);
        sgd.step();
        write_csv_row(csv, step, W, NUMEL);
    }
    csv.close();
    std::cout << "  Wrote " << csv_path << std::endl;
}


int main() {
    std::string base = "Tests/OptimizerTests/";

    std::cout << "================================================================" << std::endl;
    std::cout << "  Long-Horizon Drift Test — 500 steps, non-uniform gradients" << std::endl;
    std::cout << "================================================================" << std::endl;

    // CPU tests
    test_adam(false,  base + "drift_adam_cpu.csv");
    test_adamw(false, base + "drift_adamw_cpu.csv");
    test_sgd(false,   base + "drift_sgd_cpu.csv");

    // GPU tests
    test_adam(true,   base + "drift_adam_gpu.csv");
    test_adamw(true,  base + "drift_adamw_gpu.csv");

    std::cout << "\n================================================================" << std::endl;
    std::cout << "  Done. Now run:" << std::endl;
    std::cout << "    cd Tests/OptimizerTests" << std::endl;
    std::cout << "    python3 test_long_horizon_drift.py" << std::endl;
    std::cout << "================================================================" << std::endl;

    return 0;
}

File: DTensor/Tensor-Implementations/Tests/OptimizerTests/test_long_horizon_drift.py
#!/usr/bin/env python3
"""
test_long_horizon_drift.py

Long-horizon drift test: 500 steps with non-uniform gradients.
Runs PyTorch reference for Adam, AdamW, and SGD, then automatically
compares against C++ CSV output.

Usage:
  1. First run the C++ test:
       make run-snippet FILE=Tests/OptimizerTests/test_long_horizon_drift.cpp
  2. Then run this script from the OptimizerTests directory:
       cd Tests/OptimizerTests && python3 test_long_horizon_drift.py
"""
import torch
import numpy as np
import csv
import os
import sys


# ============================================================================
# Deterministic LCG PRNG — must match C++ exactly
# ============================================================================
def lcg_float(seed: int) -> float:
    """Returns a float in [-1.0, 1.0) given a seed. Matches C++ lcg_float()."""
    MOD = 2**64
    A = 6364136223846793005
    C = 1442695040888963407

    state = seed % MOD
    state = (A * state + C) % MOD
    state = (A * state + C) % MOD

    # Extract top 31 bits (matching C++ "state >> 33")
    bits = (state >> 33) & 0x7FFFFFFF
    return (float(bits) / float(0x7FFFFFFF)) * 2.0 - 1.0


def make_gradient(step: int, numel: int, device: torch.device) -> torch.Tensor:
    """Generate gradient tensor with deterministic non-uniform values."""
    values = []
    for i in range(numel):
        seed = step * 1000 + i
        values.append(lcg_float(seed))
    grad = torch.tensor(values, dtype=torch.float32, device=device).reshape(4, 4)
    return grad


# ============================================================================
# CSV utilities
# ============================================================================
def write_csv(path: str, steps_data: list, numel: int):
    """Write per-step parameter values to CSV."""
    with open(path, 'w', newline='') as f:
        writer = csv.writer(f)
        header = ['step'] + [f'p{i}' for i in range(numel)]
        writer.writerow(header)
        for step, values in steps_data:
            row = [step] + [f'{v:.8f}' for v in values]
            writer.writerow(row)


def read_csv(path: str):
    """Read CSV and return dict: step -> list of float values."""
    data = {}
    with open(path, 'r') as f:
        reader = csv.reader(f)
        header = next(reader)  # skip header
        for row in reader:
            step = int(row[0])
            values = [float(x) for x in row[1:]]
            data[step] = values
    return data


# ============================================================================
# PyTorch reference runners
# ============================================================================
NUM_STEPS = 5000
NUMEL = 16  # 4x4


def make_initial_weights(device: torch.device) -> torch.Tensor:
    """Create initial weights [0.1, 0.2, ..., 1.6] matching C++."""
    W = torch.zeros(4, 4, dtype=torch.float32, requires_grad=True, device=device)
    with torch.no_grad():
        for i in range(NUMEL):
            W.view(-1)[i] = 0.1 * (i + 1)
    return W


def run_adam_pytorch(device_str: str, csv_path: str):
    """Run PyTorch Adam for NUM_STEPS steps."""
    device = torch.device(device_str)
    print(f"\n=== PyTorch Adam ({device_str.upper()}) — {NUM_STEPS} steps ===")

    W = make_initial_weights(device)
    optimizer = torch.optim.Adam([W], lr=0.001, betas=(0.9, 0.999), eps=1e-8,
                                  weight_decay=0.01)

    steps_data = []
    # Step 0: initial state
    steps_data.append((0, W.detach().cpu().flatten().tolist()))

    for step in range(1, NUM_STEPS + 1):
        optimizer.zero_grad()
        W.grad = make_gradient(step, NUMEL, device)
        optimizer.step()
        steps_data.append((step, W.detach().cpu().flatten().tolist()))

    write_csv(csv_path, steps_data, NUMEL)
    print(f"  Wrote {csv_path}")


def run_adamw_pytorch(device_str: str, csv_path: str):
    """Run PyTorch AdamW for NUM_STEPS steps."""
    device = torch.device(device_str)
    print(f"\n=== PyTorch AdamW ({device_str.upper()}) — {NUM_STEPS} steps ===")

    W = make_initial_weights(device)
    optimizer = torch.optim.AdamW([W], lr=0.001, betas=(0.9, 0.999), eps=1e-8,
                                   weight_decay=0.01)

    steps_data = []
    steps_data.append((0, W.detach().cpu().flatten().tolist()))

    for step in range(1, NUM_STEPS + 1):
        optimizer.zero_grad()
        W.grad = make_gradient(step, NUMEL, device)
        optimizer.step()
        steps_data.append((step, W.detach().cpu().flatten().tolist()))

    write_csv(csv_path, steps_data, NUMEL)
    print(f"  Wrote {csv_path}")


def run_sgd_pytorch(device_str: str, csv_path: str):
    """Run PyTorch SGD (momentum + weight_decay) for NUM_STEPS steps."""
    device = torch.device(device_str)
    print(f"\n=== PyTorch SGD+Momentum+WD ({device_str.upper()}) — {NUM_STEPS} steps ===")

    W = make_initial_weights(device)
    optimizer = torch.optim.SGD([W], lr=0.01, momentum=0.9, weight_decay=0.01)

    steps_data = []
    steps_data.append((0, W.detach().cpu().flatten().tolist()))

    for step in range(1, NUM_STEPS + 1):
        optimizer.zero_grad()
        W.grad = make_gradient(step, NUMEL, device)
        optimizer.step()
        steps_data.append((step, W.detach().cpu().flatten().tolist()))

    write_csv(csv_path, steps_data, NUMEL)
    print(f"  Wrote {csv_path}")


# ============================================================================
# Comparison
# ============================================================================
def compare_csvs(own_path: str, pytorch_path: str, name: str, tolerance: float = 1e-5):
    """Compare two CSV files element-wise and report max deviation."""
    if not os.path.exists(own_path):
        print(f"  [SKIP] {name}: C++ CSV not found at {own_path}")
        print(f"         Run the C++ test first!")
        return False

    own_data = read_csv(own_path)
    pytorch_data = read_csv(pytorch_path)

    max_dev = 0.0
    worst_step = 0
    worst_elem = 0
    num_steps_checked = 0

    for step in sorted(pytorch_data.keys()):
        if step not in own_data:
            print(f"  [FAIL] {name}: Step {step} missing from C++ output")
            return False

        own_vals = own_data[step]
        pt_vals = pytorch_data[step]

        for i in range(len(pt_vals)):
            dev = abs(own_vals[i] - pt_vals[i])
            if dev > max_dev:
                max_dev = dev
                worst_step = step
                worst_elem = i
        num_steps_checked += 1

    passed = max_dev < tolerance

    print(f"\n  --- {name} ---")
    print(f"  Steps compared : {num_steps_checked}")
    print(f"  Max deviation  : {max_dev:.2e}")
    print(f"  Worst at       : step={worst_step}, element=p{worst_elem}")
    if passed:
        print(f"  Result         : [PASS] (tolerance={tolerance:.0e})")
    else:
        # Show the actual values at the worst point
        own_val = own_data[worst_step][worst_elem]
        pt_val = pytorch_data[worst_step][worst_elem]
        print(f"  OwnTensor value: {own_val:.8f}")
        print(f"  PyTorch value  : {pt_val:.8f}")
        print(f"  Result         : [FAIL] deviation {max_dev:.2e} > tolerance {tolerance:.0e}")

    return passed


# ============================================================================
# Main
# ============================================================================
def main():
    print("=" * 70)
    print("  Long-Horizon Drift Test — PyTorch Reference + Comparison")
    print("=" * 70)

    cuda_available = torch.cuda.is_available()
    if cuda_available:
        print(f"CUDA available: {torch.cuda.get_device_name(0)}")
    else:
        print("CUDA not available, GPU tests will be skipped")

    # ----- Run PyTorch references -----
    print("\n" + "-" * 40)
    print("Running PyTorch references...")
    print("-" * 40)

    run_adam_pytorch("cpu",  "drift_adam_cpu_pytorch.csv")
    run_adamw_pytorch("cpu", "drift_adamw_cpu_pytorch.csv")
    run_sgd_pytorch("cpu",   "drift_sgd_cpu_pytorch.csv")

    if cuda_available:
        run_adam_pytorch("cuda",  "drift_adam_gpu_pytorch.csv")
        run_adamw_pytorch("cuda", "drift_adamw_gpu_pytorch.csv")

    # ----- Compare against C++ output -----
    print("\n" + "=" * 70)
    print("  Comparing OwnTensor vs PyTorch")
    print("=" * 70)

    results = []

    results.append(compare_csvs("drift_adam_cpu.csv",  "drift_adam_cpu_pytorch.csv",  "Adam (CPU)"))
    results.append(compare_csvs("drift_adamw_cpu.csv", "drift_adamw_cpu_pytorch.csv", "AdamW (CPU)"))
    results.append(compare_csvs("drift_sgd_cpu.csv",   "drift_sgd_cpu_pytorch.csv",   "SGD (CPU)"))

    if cuda_available:
        results.append(compare_csvs("drift_adam_gpu.csv",  "drift_adam_gpu_pytorch.csv",  "Adam (GPU)"))
        results.append(compare_csvs("drift_adamw_gpu.csv", "drift_adamw_gpu_pytorch.csv", "AdamW (GPU)"))

    # ----- Summary -----
    print("\n" + "=" * 70)
    valid_results = [r for r in results if r is not None]
    passed = sum(1 for r in valid_results if r)
    total = len(valid_results)
    skipped = len(results) - total

    print(f"  Overall: {passed}/{total} PASSED", end="")
    if skipped > 0:
        print(f", {skipped} SKIPPED", end="")
    print()

    if passed == total and total > 0:
        print("All optimizer drift tests PASSED!")
    else:
        print("Some tests FAILED or were skipped.")
    print("=" * 70)

    sys.exit(0 if passed == total else 1)


if __name__ == "__main__":
    main()
